### PR TITLE
2247: When pr's headHash is missing, stop further processing

### PR DESCRIPTION
**Pull request by zhaosongzs**
Kevin Rushforth found this GitLab issue few years ago.

Sometimes, pull requests in GitLab would be stuck. When it happens, it shows the number of commits as "0". Clicking on the "Changes" tab shows no diffs, and displays a warning banner: "Something went wrong on our end. Please try again!" This requires action on user's part to get them "unstuck".

In Skara side, when it happens, pull request bot would not be able to get the headHash of this pull request and then throw an exception and trigger an other round of CheckWorkItem. If the user doesn't fix it, skara bot will continuously work on the pull request.

To resolve it, we should try to make pull request bot be able to identify this situation and provide some guidance to the users.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Issue
 * [SKARA-2247](https://bugs.openjdk.org/browse/SKARA-2247): When pr's headHash is missing, stop further processing (**Bug** - P4)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/skara.git pull/1645/head:pull/1645` \
`$ git checkout pull/1645`

Update a local copy of the PR: \
`$ git checkout pull/1645` \
`$ git pull https://git.openjdk.org/skara.git pull/1645/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1645`

View PR using the GUI difftool: \
`$ git pr show -t 1645`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/skara/pull/1645.diff">https://git.openjdk.org/skara/pull/1645.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/skara/pull/1645#issuecomment-2083773562)

### CODE REVIEW

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckWorkItem.java [hunk 1]
```diff
@@ -430,6 +430,19 @@ public Collection<WorkItem> prRun(ScratchArea scratchArea) {
...
 
+        if (pr.headHash().hex() == null) {
+            String text = "The head hash of this pull request is missing.\n" +
+                    "Until this is resolved, this pull request cannot be processed.\n" +
+                    "Please try to use the following command to push an empty commit.\n" +
+                    "```bash\n" +
+                    "$ git commit --allow-empty -m \"Empty commit\"\n" +
+                    "$ git push\n" +
+                    "```\n" +
+                    "If the issue still exists, please notify skara admins.";
```
Comment (erikj79) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckWorkItem.java:441]:
```suggestion
                    "If the issue still exists, please notify Skara admins.";
```

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckWorkItem.java [hunk 2]
```diff
@@ -430,6 +430,19 @@ public Collection<WorkItem> prRun(ScratchArea scratchArea) {
         var comments = prComments();
         comments = postPlaceholderForReadyComment(comments);
 
+        if (pr.headHash().hex() == null) {
+            String text = "The head hash of this pull request is missing.\n" +
+                    "Until this is resolved, this pull request cannot be processed.\n" +
+                    "Please try to use the following command to push an empty commit.\n" +
```
Comment (erikj79) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckWorkItem.java:436]:
```suggestion
                    "This is likely caused by a caching problem in the server and can usually be worked around by pushing another commit to the pull request branch. The commit can be empty. Example:\n" +
```
I also recommend removing all the explicit single line breaks. Single linebreaks will not be honored when formatting anyway and we don't want these sentences in separate paragraphs.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckWorkItem.java [hunk 3]
```diff
@@ -430,6 +430,20 @@ public Collection<WorkItem> prRun(ScratchArea scratchArea) {
         var comments = prComments();
         comments = postPlaceholderForReadyComment(comments);
 
+        if (pr.headHash().hex() == null) {
+            String text = "The head hash of this pull request is missing." +
```
Comment (erikj79) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckWorkItem.java:434]:
```suggestion
            String text = "The head hash of this pull request is missing. " +
```

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckWorkItem.java [hunk 4]
```diff
@@ -430,6 +430,20 @@ public Collection<WorkItem> prRun(ScratchArea scratchArea) {
...
+                    "Until this is resolved, this pull request cannot be processed." +
+                    "This is likely caused by a caching problem in the server and can usually be worked around by pushing another commit to the pull request branch. " +
```
Comment (erikj79) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckWorkItem.java:436]: Probably break this line a bit.

**Review by erikj79 — APPROVED** [review]

**Comment by zhaosongzs** [comment]
Thanks!
/integrate

**Pull request closed without merging**